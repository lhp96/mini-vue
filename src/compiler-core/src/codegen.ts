import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  CREATE_TEXT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;
  genFunctionPreamble(ast, context);
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}){`);
  push(`return `);

  genNode(ast.codegenNode, context);
  push(`}`);
  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  const VueBinging = "Vue";
  // 处理需要用到的所有 import
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`;
  if (ast.helpers.length) {
    push(`const { ${ast.helpers.map(aliasHelper).join(",")} } = ${VueBinging}`);
  }
  push(`\nreturn `);
}

function genNode(node: any, context: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node: any, context: any) {
  const { push, helper } = context;
  const { children } = node;
  push(`${helper(CREATE_TEXT_VNODE)}(`);
  children.forEach((child) => {
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  });
  push(`)`);
}

function genElement(node: any, context: any) {
  const { push, helper } = context;
  const { tag, props, children } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // genChildren
  genNodeList(genNullable([tag, props, children]), context);
  push(`)`);
}

function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node: any = nodes[i];
    if (isString(node)) {
      // 处理空值 props / children
      push(node);
    } else {
      // 处理 children
      if (Array.isArray(node) && node.length) {
        if (node.length === 1 && node[0].type !== NodeTypes.ELEMENT) {
          genNode(node[0], context);
        } else {
          push(`[`);
          for (let j = 0; j < node.length; j++) {
            const child = node[j];
            if (child && child.codegenNode) {
              genNode(child.codegenNode, context);
            } else {
              genNode(child, context);
            }
            if (j < node.length - 1) {
              push(", ");
            }
          }
          push(`]`);
        }
        return;
      } else {
        // console.log("进入 子genNode --- 应该不会再进入");
        genNode(node, context);
      }
    }

    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}
// 处理没有传入的数据变为 "null", eg-> props:{}, children:[]
function genNullable(args: any) {
  return args.map((arg) =>
    Array.isArray(arg) ? (arg.length ? arg : "null") : arg || "null"
  );
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  const to_display = helper(TO_DISPLAY_STRING);
  push(`${to_display}(`);
  genNode(node.content, context);
  push(`)`);
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function createCodegenContext(): any {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}
