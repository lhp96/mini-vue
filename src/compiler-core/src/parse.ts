import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context = createParseContext(content);
  return createRoot(parseChildren(context));
}
function parseChildren(context) {
  const nodes: any = [];
  let node;
  let s = context.source;
  if (s.startsWith("{{")) {
    node = parseInterpolation(context);
  } else if (s[0] === "<") {
    if (/[a-z]/i.test(s[1])) {
      node = parseElement(context);
    }
  }
  if (!node) {
    node = parseText(context);
  }
  nodes.push(node);

  return nodes;
}

function parseText(context: any) {
  const content = parseTextData(context, context.source.length);
  return {
    type: NodeTypes.TEXT,
    tag: content,
  };
}

function parseTextData(context: any, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}

function parseElement(context: any) {
  // 1. 解析 tag   2.删除处理完成的string
  const element = parseTag(context, TagType.Start);
  parseTag(context, TagType.End);
  return element;
}

function parseTag(context: any, tagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  advanceBy(context, match[0].length);
  advanceBy(context, 1);
  if (tagType === TagType.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  // {{message}}
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  // slice(start, end)
  advanceBy(context, openDelimiter.length);
  const rawContentLength = closeIndex - openDelimiter.length;
  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();
  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}
function advanceBy(context, length) {
  context.source = context.source.slice(length);
}
function createRoot(children) {
  return {
    children,
  };
}
function createParseContext(content: string): any {
  return {
    source: content,
  };
}
