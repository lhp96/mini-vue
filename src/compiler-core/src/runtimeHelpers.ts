export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREATE_ELEMENT_VNODE = Symbol("createElemetVNode");

export const helperMapName = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [CREATE_ELEMENT_VNODE]: "createElemetVNode",
};
