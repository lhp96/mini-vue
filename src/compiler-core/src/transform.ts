export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);

  createRootCodegen(root);
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0];
}

function traverseNode(node: any, context: any) {
  if (!node) return;
  const transforms = context.nodeTransforms;
  for (let i = 0; i < transforms.length; i++) {
    transforms[i](node);
  }

  traverseChildren(node, context);
}

function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      traverseNode(children[i], context);
    }
  }
}

function createTransformContext(root: any, options: any): any {
  return {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };
}
