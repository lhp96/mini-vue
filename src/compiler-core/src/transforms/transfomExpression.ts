import { NodeTypes } from "../ast";

export function transformExpression(node) {
  console.log(node);
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}
