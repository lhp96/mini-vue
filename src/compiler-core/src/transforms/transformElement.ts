import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    context.helper(CREATE_ELEMENT_VNODE);
  }
}
