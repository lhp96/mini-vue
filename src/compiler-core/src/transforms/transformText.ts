import { NodeTypes } from "../ast";
import { CREATE_TEXT_VNODE } from "../runtimeHelpers";
import { isText } from "../utils";

export function transformText(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const { children } = node;
      let currentContainer;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j];
            if (isText(next)) {
              if (!currentContainer) {
                context.helper(CREATE_TEXT_VNODE);
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child],
                };
              }
              currentContainer.children.push(" + ");
              currentContainer.children.push(next);
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = null;
              break;
            }
          }
        }
      }
    };
  }
}
