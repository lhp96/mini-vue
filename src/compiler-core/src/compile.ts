import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformExpression } from "./transforms/transfomExpression";
import { transformElement } from "./transforms/transformElement";
import { transformText } from "./transforms/transformText";

export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  return generate(ast);
}
