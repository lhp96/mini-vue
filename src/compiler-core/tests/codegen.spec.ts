import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transfomExpression";
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";
describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);

    // 快照(string)
    // 1. 抓bug
    // 2. 有意()
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{   msg  }}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    // console.log(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse(`<div><p></p></div>`);
    transform(ast, {
      nodeTransforms: [transformElement],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element & text & interpolation", () => {
    const ast: any = baseParse(`<div>hi,{{  message }}</div>`);
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
