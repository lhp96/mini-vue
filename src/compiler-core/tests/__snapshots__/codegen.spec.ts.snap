// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen complex template 1`] = `
"const { toDisplayString: _toDisplayString,createElementVNode: _createElementVNode,createTextVNode: _createTextVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, [_createTextVNode('hi,' + _toDisplayString(_ctx.count) + '---- written by lhp96'), _createElementVNode('h1', null, 'have a nice Day'), _createElementVNode('p', null, '2022.9.28')])}"
`;

exports[`codegen element & text & interpolation 1`] = `
"const { toDisplayString: _toDisplayString,createTextVNode: _createTextVNode,createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, _createTextVNode('hi,' + _toDisplayString(_ctx.message)))}"
`;

exports[`codegen element 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, [_createElementVNode('p', null, null)])}"
`;

exports[`codegen element Nested 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, [_createElementVNode('p', null, null), _createElementVNode('t', null, null), _createElementVNode('p', null, null)])}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.msg)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;
